solr/core/src/java/org/apache/solr/servlet/RateLimitManager.java

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * This class is responsible for managing rate limiting per request type. Rate limiters
+ * can be registered with this class against a corresponding type. There can be only one
+ * rate limiter associated with a request type.
+ *
+ * The actual rate limiting and the limits should be implemented in the corresponding RequestRateLimiter
+ * implementation. RateLimitManager is responsible for the orchestration but not the specifics of how the
+ * rate limiting is being done for a specific request type.
+ */
+public class RateLimitManager {
+  public final static int DEFAULT_CONCURRENT_REQUESTS= 10;
+  public final static long DEFAULT_EXPIRATION_TIME_INMS = 300;
+  public final static long DEFAULT_SLOT_ACQUISITION_TIMEOUT_MS = -1;
+
+  private final Map<String, RequestRateLimiter> requestRateLimiterMap;
+
+  public RateLimitManager() {
+    this.requestRateLimiterMap = new HashMap<String, RequestRateLimiter>();
+  }
+
+  // Handles an incoming request. The main orchestration code path, this method will
+  // identify which (if any) rate limiter can handle this request. Internal requests will not be
+  // rate limited
+  // Returns true if request is accepted for processing, false if it should be rejected
+
+  // NOTE: It is upto specific rate limiter implementation to handle queuing of rejected requests.
+  public boolean handleRequest(HttpServletRequest request) throws InterruptedException {
+    String requestContext = request.getHeader(SOLR_REQUEST_CONTEXT_PARAM);
+    String typeOfRequest = request.getHeader(SOLR_REQUEST_TYPE_PARAM);
+
+    if (typeOfRequest == null) {
+      // Cannot determine if this request should be throttled
+      return true;
+    }
+
+    // Do not throttle internal requests
+    if (requestContext != null && requestContext.equals(SolrRequest.SolrClientContext.SERVER.toString())) {
+      return true;
+    }
+
+    RequestRateLimiter requestRateLimiter = requestRateLimiterMap.get(typeOfRequest);
+
+    if (requestRateLimiter == null) {
+      // No request rate limiter for this request type
+      return true;
+    }
+
+    return requestRateLimiter.handleRequest(request);
+  }
+
+  // Resume a pending request from one of the registered rate limiters.
+  // The current model is round robin -- iterate over the list and get a pending request and resume it.
+
+  // TODO: This should be a priority queue based model
+  public void resumePendingRequest(HttpServletRequest request) {
+    String typeOfRequest = request.getHeader(SOLR_REQUEST_TYPE_PARAM);
+
+    RequestRateLimiter previousRequestRateLimiter = requestRateLimiterMap.get(typeOfRequest);
+
+    if (previousRequestRateLimiter == null) {
+      // No rate limiter for this request type
+      return;
+    }
+
+    // Give preference to the previous request's rate limiter
+    if (previousRequestRateLimiter.resumePendingOperation()) {
+      return;
+    }
+
+    for (Map.Entry<String, RequestRateLimiter> currentEntry : requestRateLimiterMap.entrySet()) {
```

Comment:
I'm confused as to why this exists. If a query request completes, why does that mean that we have room for an update request now?

Reply:
This is work stealing -- a thread first tries to get work from its own request rate limiter and if not present, let the thread be used for a different request type. This is to allow smoothening out of skews. Added a flag to enable or disable this behaviour.